analytics/pubstack/pubstack_module.go

```diff
@@ -0,0 +1,301 @@
+		return
+	}
+
+	p.logObject(setUID, payload)
+}
+
+func (p *PubstackModule) LogCookieSyncObject(cso *analytics.CookieSyncObject) {
+	if !p.isFeatureEnable(cookieSync) {
+		return
+	}
+
+	// serialize event
+	payload, err := helpers.JsonifyCookieSync(cso, p.scope)
+	if err != nil {
+		glog.Warning("[pubstack] Cannot serialize video")
+		return
+	}
+
+	p.logObject(cookieSync, payload)
+}
+
+func (p *PubstackModule) LogAmpObject(ao *analytics.AmpObject) {
+	if !p.isFeatureEnable(amp) {
+		return
+	}
+
+	// serialize event
+	payload, err := helpers.JsonifyAmpObject(ao, p.scope)
+	if err != nil {
+		glog.Warning("[pubstack] Cannot serialize video")
+		return
+	}
+
+	p.logObject(amp, payload)
+}
+
+func (p *PubstackModule) fetchAndUpdateConfig(refreshDelay time.Duration, endCh chan os.Signal) {
+	tick := time.NewTicker(refreshDelay)
+
+	for {
+		select {
+		case <-tick.C:
+			config, err := fetchConfig(p.cfg.ScopeId, p.cfg.Endpoint)
+			if err != nil {
+				glog.Errorf("[pubstack] Fail to fetch remote configuration: %v", err)
+				continue
+			}
+			p.configCh <- config
+		case <-endCh:
+			return
+		}
+	}
+}
+
+func (p *PubstackModule) setup() {
+	config, err := fetchConfig(p.cfg.ScopeId, p.cfg.Endpoint)
+	if err != nil {
+		glog.Errorf("[pubstack] Fail to fetch remote configuration: %v", err)
+		config = p.cfg
+	}
+	p.configCh <- config
+}
+
+func (p *PubstackModule) start(refreshDelay time.Duration, endCh chan os.Signal) {
+
+	// update periodically the config
+	go p.fetchAndUpdateConfig(refreshDelay, endCh)
+
+	for {
+		select {
+		case config := <-p.configCh:
+			p.mux.Lock()
+			p.cfg = config
+			p.initEventChannels()
+			p.mux.Unlock()
+			glog.Infof("[pubstack] Updating config: %v", p.cfg)
+		case <-endCh:
+			return
+		}
+	}
+
+}
+
+func fetchConfig(scope string, intake string) (*Configuration, error) {
+	u, err := url.Parse(intake)
+	if err != nil {
+		return nil, err
+	}
+
+	u.Path = path.Join(u.Path, "bootstrap")
+	q, _ := url.ParseQuery(u.RawQuery)
+
+	q.Add("scopeId", scope)
+	u.RawQuery = q.Encode()
+
+	res, err := clients.GetDefaultHttpInstance().Get(u.String())
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Body.Close()
+	c := Configuration{}
+	err = json.NewDecoder(res.Body).Decode(&c)
+	if err != nil {
+		return nil, err
+	}
+	return &c, nil
+}
+
+func newBufferConfig(count int, size, duration string) (*bufferConfig, error) {
+	pDuration, err := time.ParseDuration(duration)
+	if err != nil {
+		return nil, err
+	}
+	pSize, err := units.FromHumanSize(size)
+	if err != nil {
+		return nil, err
+	}
+	return &bufferConfig{
+		pDuration,
+		int64(count),
+		pSize,
+	}, nil
+}
+
+func buildEndpointSender(baseUrl string, module string) eventchannel.Sender {
+	endpoint, err := url.Parse(baseUrl)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	endpoint.Path = path.Join(endpoint.Path, "intake", module)
+	return eventchannel.NewHttpSender(endpoint.String())
+}
+
+func (p *PubstackModule) initEventChannels() {
+	if p.isFeatureEnable(amp) {
```

Comment:
Since `initEventChannels()` is called each time the config is refreshed, don't we have a resource/memory leak. Say for example the config is set to refresh hourly and an instance of PBS is left running for a week. That would result in 168 amp event channels, each with its own go routine listening for messages. Of course only one instance would receive any messages other than the periodic flush buffers timer.

Reply:
Good catch! I've refactored that part, and closed the channels

Reply:
I believe you have created a race condition here. By deleting the old channels first, you create a temporary condition where the map `p.eventChannels` will not have a value for a feature, like amp, while `p.isFeatureEnabled` is still true. You use only `p.isFeatureEnabled` to guard against calling `p.eventChannels[feature].Push(payload)` on a feature that is not present in `p.eventChannels`. You could put the existence check in `p.logObject`, but that would still leave a tiny window for race conditions. Better to save the pointers to the old channels, and then close the old channel after the new channel has been slotted into the map.

This may still not close all the race condition possibility upon reflection. Updating a map with a new value may not be an atomic action, certainly not for a struct where values need to be copied in. In the case of a pointer, it is less clear, as in go pointers are not a simple address, but also contain metadata about the object they are pointing to. You may need something like this in `logObject` to close out the possibility completely:
```
ch, ok := p.eventChannels[feature]
if ok && ch != nil {
    ch.Push(payload)
}
```
I would hope go is smart enough to not signal `ok` on pulling a map value if the map is mid update, but I have not researched this point before.

Reply:
Perhaps the safest approach is to add an `isActive bool` property to `PubstackModule`, set it to false at the start of `setEventChannels()`, and not set it back to true until the new channels have been created. Then have a line to have `logObject()` return before doing anything if this flag is false. Alternatively you could have `setEventChannels()` set the same `p.mux.Lock()` to lock out `logObject()` until the channels are stable again. You should also consider locking your module until the initial config load is done, you don't want the analytics to start flowing before the module is up and running. ... Reviewing again, it seems you don't return the pubstack object until the initial config load is done, so that is probably not an issue.

Reply:
Thanks for your reply, I will check it today

Reply:
*Disclaimer, I've rebased the branch on the lastest master and squash all previous commits to clean the PR.*

What do you think moving the Mutex in each `Log*Object` methods? Since, we can't use generics, it doesn't make a lot of sense to have the `log object` method jsut for pushing the events.

I've also refactored the `setEventChannels` to `configure`. The goal is to create an "atomic" operation (with the Mutex) to update the module. So, during the update, the `Log*Object` methods are blocked/synced. 

I think it removes the race condition.

It's been a long time I didn't code in go, so I have to switch my mind from an eventloop model (js) to a multi thread app :sunglasses:

Reply:
Yes, that should do it. The race condition is on losing the channel, so at the time of the `p.eventChannels[].Push(payload)` the channel must exist. You do run into the possibility of the shutdown being processed while there are still payloads in the queue, so you could potentially lose a few events. But that should be an insignificant fraction of the total, especially if you have a nice long delay between config refresh events.